### DISCUSSION TITLE
Exception On SaveChanges when Audit.Net Entity framework have High-Level Interceptions

### DISCUSSION THREAD

**Discussion by DiegoQuirosT** [thepirat000/Audit.NET#679] [Q&A] [answered]
Hello, I have encountered a small problem. I am trying to audit my entities, but when I attempt to call SaveChanges with the interceptor implemented, I receive this error: System.MissingMethodException: 'Constructor on type 'Appia.Store.AppDBC' not found.'
However, I do have a constructor. Here is my DbContext code and my configuration:

 `using Audit.EntityFramework;
using Microsoft.AspNetCore.DataProtection.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore;

namespace Appia.Store;

public class AppDBC : DbContext, IDataProtectionKeyContext
{
    public DbSet<DataProtectionKey> DataProtectionKeys { get; set; } = null!;
    public DbSet<Organization> Organizations { get; set; }
    public DbSet<License> Licenses { get; set; }
    public DbSet<Event> Events { get; set; }

    private readonly DbContextHelper _helper = new DbContextHelper();
    private readonly IAuditDbContext _auditContext;

    public AppDBC(DbContextOptions<AppDBC> options) : base(options)
    {
        _auditContext = new DefaultAuditContext(this);
        _helper.SetConfig(_auditContext);
    }

    public override int SaveChanges(bool acceptAllChangesOnSuccess)
    {
        return _helper.SaveChanges(_auditContext, () => base.SaveChanges(acceptAllChangesOnSuccess)); //error throws here
    }

    public override async Task<int> SaveChangesAsync(bool acceptAllChangesOnSuccess, CancellationToken cancellationToken = default(CancellationToken))
    {
        return await _helper.SaveChangesAsync(_auditContext, () => base.SaveChangesAsync(acceptAllChangesOnSuccess, cancellationToken));
    }

    protected override void OnModelCreating( ModelBuilder builder )
    {
        builder.ApplyConfiguration( new EfConfigLicense() );
        builder.ApplyConfiguration( new EfConfigOrganization() );
        builder.ApplyConfiguration( new EfConfigEvent() );
    }

    public void AuditInfo(string User, string UserId, string Comment, string Email)
    {
        _auditContext.ExtraFields[nameof(CheckIn.ByUser)] = User;
        _auditContext.ExtraFields[nameof(CheckIn.ByUserId)] = UserId;
        _auditContext.ExtraFields[nameof(CheckIn.Comments)] = Comment;
        _auditContext.ExtraFields[nameof(CheckIn.ByEmail)] = Email;
    }
}

        Audit.Core.Configuration.Setup()
            .UseEntityFramework(ef => ef
            .UseDbContext<AppDBC>()
            .DisposeDbContext(true)
            .AuditTypeMapper(t => typeof(Event))
            .AuditEntityAction<Event>((ev, entry, entity) =>
            {
                entity.EntityId = entry.PrimaryKey.First().Value.ToString();
                entity.Schema = entry.Schema;
                entity.Table = entry.Table;
                entity.Name = entry.Name;
                entity.Action = entry.Action;
                entity.Entity = entry.Entity;
                entity.Changes = entry.Changes;
                entity.Valid = entry.Valid;
                entity.ValidationResults = entry.ValidationResults;
                entity.EntityType = entry.EntityType;
                entity.EntityTypeName = entry.EntityType.Name;
                entity.Created = new CheckIn
                {
                    At = DateTime.Now,
                    WithApp = "Appia.Store",
                    Comments = entry.GetEntry().CurrentValues.GetValue<string>(nameof(CheckIn.Comments)),
                    ByEmail = entry.GetEntry().CurrentValues.GetValue<string>(nameof(CheckIn.ByEmail)),
                    ByUser = entry.GetEntry().CurrentValues.GetValue<string>(nameof(CheckIn.ByUser)),
                    ByUserId = entry.GetEntry().CurrentValues.GetValue<string>(nameof(CheckIn.ByUserId)),
                };
                entity.Id = Guid.NewGuid().ToString();
                entity.Value = entry.ToJson();
            })

        );

**Comment by thepirat000** [comment, marked as answer]
The constructor of your DbContext requires one parameter of type `DbContextOptions<AppDBC>`.

Try changing this line 

```c#
.UseDbContext<AppDBC>()
```

To something like this:

```c#
.UseDbContext<AppDBC>(new DbContextOptionsBuilder<AppDBC>().UseSqlServer("YourConnectionString").Options)
```

**Reply by DiegoQuirosT** [reply]
It worked! I didn’t know that it was indeed a new DbContext, but I just figured out how to use the original.

Thank you!

**Reply by thepirat000** [reply]
ok. Have in mind that if you use the same DbContext instance for the audit, whenever the SaveChanges fails, the DbContext instance will be set to a faulted state and the audit will also fail to save.

https://github.com/thepirat000/Audit.NET/issues/589#issuecomment-1481683243
